# Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
#
# NVIDIA CORPORATION, its affiliates and licensors retain all intellectual
# property and proprietary rights in and to this material, related
# documentation and any modifications thereto. Any use, reproduction,
# disclosure or distribution of this material and related documentation
# without an express license agreement from NVIDIA CORPORATION or
# its affiliates is strictly prohibited.

# Behavior Tree Configuration Parameters
behavior_tree_params:
  # Multi-object pick and place behavior tree settings
  multi_object_pick_and_place:

    assign_object_name:
      service_name: 'assign_name_to_object'

    attach_object:
      action_server_name: 'attach_object'
      fallback_radius: 0.09
      shape: 'CUBOID'
      scale: [0.1, 0.1, 0.2]
      gripper_frame: 'gripper_frame'
      grasp_frame: 'grasp_frame'

    close_gripper:
      gripper_action_name: '/robotiq_gripper_controller/gripper_cmd'
      close_position: 0.55
      max_effort: 10.0

    detach_object:
      action_server_name: 'attach_object'

    detect_object:
      action_server_name: '/get_objects'
      detection_confidence_threshold: 0.5

    execute_trajectory:
      action_server_name: 'execute_trajectory'
      index: 0

    interactive_marker:
      mesh_resource_uri: 'package://isaac_manipulator_robot_description/meshes/robotiq_2f_85.obj'
      reference_frame: 'base_link'
      end_effector_frame: 'gripper_frame'
      user_confirmation_timeout: 60.0  # seconds to wait for user to adjust drop pose

    mesh_assigner:
      service_name: 'add_mesh_to_object'

    object_selector:
      action_server_name: '/get_selected_object'

    open_gripper:
      gripper_action_name: '/robotiq_gripper_controller/gripper_cmd'
      open_position: 0.0
      max_effort: 10.0

    plan_to_grasp:
      action_server_name: 'cumotion/motion_plan'
      link_name: 'base_link'
      time_dilation_factor: 0.2
      grasp_approach_offset_distance: [0.0, 0.0, -0.15]
      grasp_approach_path_constraint: [0.5, 0.5, 0.5, 0.1, 0.1, 0.0]
      retract_offset_distance: [0.0, 0.0, 0.15]
      retract_path_constraint: [0.1, 0.1, 0.1, 0.1, 0.1, 0.0]
      grasp_approach_constraint_in_goal_frame: true
      retract_constraint_in_goal_frame: false
      disable_collision_links: []
      update_planning_scene: true
      world_frame: 'world'
      enable_aabb_clearing: true
      esdf_clearing_padding: [0.05, 0.05, 0.05]

    plan_to_pose:
      action_server_name: 'cumotion/motion_plan'
      link_name: 'base_link'
      time_dilation_factor: 0.2
      update_planning_scene: true
      disable_collision_links: []
      aabb_clearing_shape: 'SPHERE'
      aabb_clearing_shape_scale: [0.1, 0.1, 0.1]
      enable_aabb_clearing: false
      esdf_clearing_padding: [0.05, 0.05, 0.05]

    pose_estimation:
      action_server_name: '/get_object_pose'
      base_frame_id: 'base_link'
      camera_frame_id: 'camera_1_color_optical_frame'  # Use 'front_stereo_camera_left' for Isaac Sim
      workspace_bounds:
        diagonal: []

    publish_static_planning_scene:
      service_name: 'publish_static_planning_scene'
      scene_file_path: ''

    read_grasp_poses:
      publish_grasp_poses: true

    retry_config:
      max_planning_retries: 5
      max_controller_retries: 2
      max_detection_retries: 3
      max_pose_estimation_retries: 3
      max_gripper_retries: 2
      max_attachment_retries: 3

    server_timeout_config:
      startup_server_timeout_sec: 120.0    # Max time to wait for servers during startup (None = wait forever)
      runtime_retry_timeout_sec: 60.0      # Max time to retry server during execution
      server_check_interval_sec: 30.0      # Interval between server checks (also determines logging frequency)

    stale_detection:
      timeout_duration: 300.0

    switch_controllers:
      arm: 
        controllers_to_activate: ['scaled_joint_trajectory_controller']
        controllers_to_deactivate: ['impedance_controller']
        strictness: 4  # FORCE_AUTO
      tool:
        controllers_to_activate: ['robotiq_activation_controller', 'robotiq_gripper_controller']
        controllers_to_deactivate: []
        strictness: 4  # FORCE_AUTO
